multidict/_multidict_c.c

```diff
@@ -0,0 +1,854 @@
+            return -1;
+        }
+        Py_DECREF(item);
+    }
+
+    Py_DECREF(iter);
+    return 0;
+}
+
+static int
+_multidict_extend(_MultiDictObject *self, PyObject *args, PyObject *kwds,
+                  const char *name, int do_add)
+{
+    // TOOD: Refactoring me. Split on little functions;
+
+    PyObject *arg        = NULL,
+             *arg_items  = NULL,
+             *kwds_items = NULL;
+
+    if (PyObject_Length(args) > 1) {
+        PyErr_Format(PyExc_TypeError,
+                     "%s takes at most 1 positional argument (%zd given)",
+                     name, PyObject_Length(args), NULL);
+        goto fail;
+    }
+
+    if (args != Py_None) {
+        arg = PyTuple_GetItem(args, 0);
+        if (!arg) {
+            goto fail;
+        }
+        Py_INCREF(arg);
+
+        // TODO: fixme multidict_base_type
+        if (MultiDict_Check(arg) && kwds == Py_None) {
+            if (do_add) {
+                // TODO: add err check
+                _multidict_append_items(
+                    self,
+                    (_MultiDictObject*)((_MultiDictObject*)arg)->impl
+                );
+            } else {
+                // TODO: add err check
+                _multidict_update_items(
+                    self,
+                    (_MultiDictObject*)((_MultiDictObject*)arg)->impl
+                );
+            }
+        } else {
+            // TODO: FIX ME!
+            // Totaly wrong!!! Rewrite this part of code!
+            // if (PyObject_HasAttrString(arg, "items")) {
+            //     arg_items = multidict_items(self);
+            // }
+
+            if (kwds != Py_None) {
+                kwds_items = PyDict_Items(kwds);
+                _multidict_list_extend(arg_items, kwds_items);
+            }
+
+            if (do_add) {
+                // TODO: add err check
+                _multidict_append_items_seq(self, arg, name);
+            } else {
+                // TODO: add err check
+                pair_list_update_from_seq(self->impl, arg);
+            }
+        }
+    } else {
+        arg = PyDict_Items(kwds);
+        if (do_add) {
+            // TODO: add err check
+            _multidict_append_items_seq(self, arg, name);
+        } else {
+            pair_list_update_from_seq(self->impl, arg);
+        }
+    }
+
+    Py_DECREF(arg);
+    Py_DECREF(arg_items);
+    Py_DECREF(kwds_items);
+    return 0;
+
+fail:
+    Py_XDECREF(arg);
+    Py_XDECREF(arg_items);
+    Py_XDECREF(kwds_items);
+    return -1;
+}
+
+static PyObject *
+multidict_getall(_MultiDictObject *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *list     = NULL,
+             *key      = NULL,
+             *_default = NULL;
+
+    static char *keywords[] = {"key", "default"};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|O:getall",
+                                     keywords, &key, &_default))
+    {
+        return NULL;
+    }
+
+    list = pair_list_get_all(self->impl, key);
```

Comment:
The same problem with setting and clearing `KeyError`

Reply:
Done if I understand right :)